Update ai to check for game end condition on Resume

bofstein opened the pull request:
Since board.js used to check for game end condition (no sets, no cards left in deck) before it drew 3 cards, if the final draw left the board with no sets, the game would never go to end.  Moving it to the ai Resume function makes it check after cards are drawn instead, which should catch all instance instead of missing that 3% edge case, since the ai resumes at the start of each play.

I tried at first to add it to the board.js file right after draw 3 but that didn't work, I think because of the time delay on the drawing.  Having it check whenever the ai resumes ensures it would check again each turn to see if it needed to end if there were no valid sets and 0 cards in deck.  

I tested by speeding up the ai and animations and playing until I hit that state and confirming it ended the game, whereas before it would just sit there forever since it hadn't ended the game and had no more cards to add.  You can tell it succeeded here because the bot has 23 sets: 81 cards - 12 on the table without any sets = 69 cards = 23 sets.

<img width="960" alt="image" src="https://github.com/bofstein/set-game/assets/23406931/106728ac-82ec-4035-b6c6-2db967e308db">

bofstein committed:
Update ai to check for game end condition on Resume

Since board.js checks for game end condition (no sets, no cards left in deck) before it draws 3 cards, if the final draw leaves the board with no sets, the game would never go to end.  Adding it to the ai Resume function makes it check after cards are drawn as well in case it needs to end

fehknt commented on js/ai.js:
```diff
@@ -67,6 +67,14 @@ export default {
         this.foundSet = false;
 
+        //have ai check each time it resumes if game should end; copied from board logic
+        if (isDev)  console.log(`ai sets found: ${this.findValidSets().length}`);
```
bring out the computation of findvalidsets into a call like let numValidSets = ....length; to avoid double computation (not needed for stock because it's a trivial lookup)

bofstein replied:
Done in latest commit

fehknt commented on js/ai.js:
```diff
@@ -67,6 +67,14 @@ export default {
 
+        //have ai check each time it resumes if game should end; copied from board logic
+        if (isDev)  console.log(`ai sets found: ${this.findValidSets().length}`);
+        if (isDev)  console.log(`ai cards left: ${deck.stock.length}`);
```
suggest making the logging a single logging line with both elements being checked in one.

bofstein replied:
Done in latest commit

fehknt commented on js/ai.js:
```diff
@@ -67,6 +67,14 @@ export default {
+        if (isDev)  console.log(`ai sets found: ${this.findValidSets().length}`);
+        if (isDev)  console.log(`ai cards left: ${deck.stock.length}`);
+        if (this.findValidSets().length === 0 && deck.stock.length === 0) {
+            setTimeout(() => game.end(), game.delay['add-cards']);
```
maybe add a game.delayEnd() function that encapsulates the timeout so that if in the future a different delay is used, it would be consistent across end calls.

bofstein replied:
Upon review and testing, it doesn't seem like this version needs a timeout since it doesn't need to wait for any card actions.  I've removed the timeout entirely in the latest commit, and am going to do another commit to remove the alternate end condition that happens in board.js since I believe checking on each ai Resume will catch all instances on its own.

fehknt reviewed (changes requested):
left some feedback

bofstein committed:
Cleanup ai.js changes

Made 3 changes in response to PR feedback:
- Created numValidSets variable so it only needs to compute that once instead of twice
- Combined console log into one line
- Removed unnecessary Timeout since it doesn't need to wait for cards

bofstein committed:
Remove end game check after points increment

The ai now checks for game end condition (no sets on board, no cards in deck) each time it resumes, which I believe will catch all instances of when the game should end, making this check after each set is made no longer necessary.

fehknt commented on js/board.js:
```diff
@@ -140,4 +135,4 @@ export default {
             deck.updateCardPos($card, newPos);
         }
     }
-}
```
nit: what changed here? can you revert this no-op line?

bofstein replied:
I can't figure out how - I didn't add/remove this myself as far as I know

fehknt commented on js/ai.js:
```diff
@@ -68,10 +68,10 @@ export default {
         this.foundSet = false;
 
         //have ai check each time it resumes if game should end; copied from board logic
```
nit: add space at start of comment.
nit: no longer true that it was copied from board logic.

bofstein replied:
Updated; now reads

// have ai check each time it resumes if game should end; this logic used to be in board.js before card draw

fehknt commented:
PR description is no longer up to date after the latest changes

bofstein commented:
> PR description is no longer up to date after the latest changes

Updated

bofstein committed:
Update ai.js comment on end game logic

Updated comment on new addition to be accurate and have a space at the star

bofstein commented:
@fehknt addressed all changes but the no-op that I can't figure out how it happened or how to revert.  Given we agree it's a no-op I'd request an approval but that's up to you if its important to track down.

fehknt reviewed (approved):
thanks for the updates! much cleaner now.
